### PR TITLE
feat: add httpc.Do & httpc.Service.Do

### PR DESCRIPTION
**Pull request by kevwan**
The final method signatures are:

```go
Do(ctx context, method, url string, data interface{}) (*http.Response, error)

type Service interface {
    Do(ctx context.Context, method, url string, data interface{}) (*http.Response, error)
    ...
}
```

The methods automatically convert `data` defined in `API` files into the `*http.Request`.

Validation is done automatically  by the `Do` methods.

### CODE REVIEW

File: core/mapping/marshaler.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+	m, ok := collector[tag]
+	if ok {
+		m[key] = val
+	} else {
+		m = map[string]interface{}{
+			key: val,
+		}
+	}
+	collector[tag] = m
```
Comment (zhoushuguang) [inline core/mapping/marshaler.go:74]: 这一行作用是什么

File: core/mapping/marshaler.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+	return nil
+}
+
+func validateOptions(value reflect.Value, opt *fieldOptions) error {
+	var found bool
+	val := fmt.Sprint(value.Interface())
+	for i := range opt.Options {
+		if opt.Options[i] == val {
+			found = true
```
Comment (zhoushuguang) [inline core/mapping/marshaler.go:125]: break?

File: core/mapping/marshaler.go (hunk 1 above)
Comment (kevwan) [reply core/mapping/marshaler.go:74]: 比如有这样的对象

```go
	type Data struct {
		Key    string `path:"key"`
		Value  int    `form:"value"`
		Header string `header:"X-Header"`
		Body   string `json:"body"`
	}
```

我需要把所有的成员按照 path, form, header, json 分类放好，等同于我们在 unmarshal 时候的逆处理

File: core/mapping/marshaler.go (hunk 2 above)
Comment (kevwan) [reply core/mapping/marshaler.go:125]: fixed.

**Review by re-dylan — APPROVED** [review]
LGTM

**Pull request merged**